File: networking_tool/network_monitoring/network_scanner.py
#!/usr/bin/env python3
from scapy.all import *
import time
import threading
import subprocess
import sys
import sqlite3
from collections import defaultdict
import yaml


# ==================== CONFIG ====================

CONF_FILE = "config/config.yaml"
with open(CONF_FILE, "r") as file:
    config = yaml.safe_load(file)

network_scanning_conf = config.get("network_scanning")
IFACE = network_scanning_conf.get("interface")
HOPPING = network_scanning_conf.get("enable_hopping")
HOP_INTERVAL = network_scanning_conf.get("hop_interval")
CHANNELS_24 = network_scanning_conf.get("ch_2_4G")
CHANNELS_5G = network_scanning_conf.get("ch_5G")
ALL_CHANNELS = CHANNELS_24 + CHANNELS_5G

db_conf = config.get("database")

DB_FILE = db_conf.get("path")
enable_persistance = db_conf.get("enable")


# ==================== DATABSE ====================

if enable_persistance:

    # Connect to SQLite DB
    conn = sqlite3.connect(DB_FILE)
    cursor = conn.cursor()

    # Create tables if not exist
    cursor.execute('''
    CREATE TABLE IF NOT EXISTS aps (
        bssid TEXT PRIMARY KEY,
        ssid TEXT,
        channel TEXT,
        freq TEXT,
        encryption TEXT,
        rssi_last INTEGER,
        rssi_max INTEGER,
        beacons_seen INTEGER,
        last_seen REAL
    )
    ''')

    cursor.execute('''
    CREATE TABLE IF NOT EXISTS clients (
        mac TEXT PRIMARY KEY,
        probes TEXT,  -- comma-separated
        rssi_last INTEGER
    )
    ''')

    cursor.execute('''
    CREATE TABLE IF NOT EXISTS associations (
        ap_bssid TEXT,
        client_mac TEXT,
        FOREIGN KEY (ap_bssid) REFERENCES aps(bssid),
        FOREIGN KEY (client_mac) REFERENCES clients(mac),
        PRIMARY KEY (ap_bssid, client_mac)
    )
    ''')
    conn.commit()


# ==================== DATA STRCUTURE ====================

aps = defaultdict(lambda: {
    "ssid": "<unknown>",
    "channel": "?",
    "freq": "?",
    "encryption": "?",
    "rssi_last": -100,
    "rssi_max": -100,
    "beacons_seen": 0,
    "last_seen": 0,
    "clients": set(),
    "probes": set()
})


# ==================== FUNCTIONS ====================

def save_to_db(bssid=None, client_mac=None):
    if bssid:
        data = aps[bssid]
        cursor.execute('''
        INSERT OR REPLACE INTO aps (bssid, ssid, channel, freq, encryption, rssi_last, rssi_max, beacons_seen, last_seen)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (bssid, data["ssid"], data["channel"], data["freq"], data["encryption"],
              data["rssi_last"], data["rssi_max"], data["beacons_seen"], data["last_seen"]))

        for cl in data["clients"]:
            cursor.execute('INSERT OR IGNORE INTO associations (ap_bssid, client_mac) VALUES (?, ?)', (bssid, cl))

    if client_mac:
        data = aps[client_mac]
        probes_str = ",".join(data["probes"])
        cursor.execute('''
        INSERT OR REPLACE INTO clients (mac, probes, rssi_last)
        VALUES (?, ?, ?)
        ''', (client_mac, probes_str, data["rssi_last"]))

    conn.commit()

def set_channel(ch):
    try:
        subprocess.check_call(["iw", "dev", IFACE, "set", "channel", str(ch)], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        print(f"[HOP] Switched to channel {ch}")
    except subprocess.CalledProcessError as e:
        print(f"[ERROR] Failed to set channel {ch}: {e}")
        sys.exit(1)

def channel_hopper():
    idx = 0
    while HOPPING:
        ch = ALL_CHANNELS[idx % len(ALL_CHANNELS)]
        set_channel(ch)
        time.sleep(HOP_INTERVAL)
        idx += 1

def freq_to_channel(freq):
    if freq is None or freq < 2400:
        return "?"
    if 2400 <= freq <= 2500:
        return (freq - 2407) // 5
    elif 5000 <= freq <= 6000:
        return (freq - 5000) // 5
    return "?"

def get_channel(pkt):
    if pkt.haslayer(RadioTap):
        rt = pkt[RadioTap]
        if hasattr(rt, 'Channel') and rt.Channel != 0:
            return rt.Channel
        if hasattr(rt, 'ChannelFrequency') and rt.ChannelFrequency != 0:
            return freq_to_channel(rt.ChannelFrequency)
    # Safer DS fallback
    elt = pkt.getlayer(Dot11Elt, ID=3)
    if elt and elt.info and len(elt.info) == 1:  # Only if exactly 1 byte
        try:
            return ord(elt.info[0])
        except:
            pass
    return "?"

def get_encryption(pkt):
    """Manual 802.11 IE parsing - more reliable than network_stats()"""
    if not pkt.haslayer(Dot11Beacon):
        return "N/A"

    # Walk through all Dot11Elt layers (chained)
    elts = pkt[Dot11Elt]
    while elts:
        if elts.ID == 48:  # RSN Information Element = WPA2/WPA3
            info = bytes(elts.info)
            if b'\x01\x00' in info:  # RSN version 1
                if b'\x02' in info or b'\x08' in info:  # WPA2-PSK or WPA3-SAE
                    return "WPA2/WPA3"
        elif elts.ID == 221:  # Vendor Specific (WPA)
            info = bytes(elts.info)
            if info.startswith(b'\x00\x50\xF2\x01'):  # Microsoft OUI + WPA
                return "WPA"
            elif info.startswith(b'\x00\x50\xF2\x04'):  # Microsoft OUI + WPA2 pre-RSN
                return "WPA2"
        elts = elts.payload.getlayer(Dot11Elt) if elts.payload else None

    # Fallback: Check Dot11 frame privacy bit (basic WEP detection)
    if pkt[Dot11].FCfield & 0x40:  # Protected frame bit
        return "WEP/PRIVACY"

    return "OPEN"


def packet_handler(pkt):
    try:
        if not pkt.haslayer(Dot11):
            return

        dot11 = pkt[Dot11]
        rssi = pkt.dBm_AntSignal if hasattr(pkt, 'dBm_AntSignal') else -100
        now = time.time()

        if pkt.haslayer(Dot11Beacon):
            bssid = dot11.addr3.upper()
            ssid = "<hidden>"
            elts = pkt[Dot11Elt]
            while elts:
                if elts.ID == 0:  # SSID IE
                    ssid = elts.info.decode('utf-8', errors='ignore').strip() or "<hidden>"
                    break
                elts = elts.payload.getlayer(Dot11Elt) if elts.payload else None

            channel = get_channel(pkt)
            freq = pkt[RadioTap].ChannelFrequency if pkt.haslayer(RadioTap) and hasattr(pkt[RadioTap], 'ChannelFrequency') else "?"

            entry = aps[bssid]
            entry["ssid"] = ssid
            entry["channel"] = channel
            entry["freq"] = freq
            entry["encryption"] = get_encryption(pkt)
            entry["rssi_last"] = rssi
            entry["rssi_max"] = max(entry["rssi_max"], rssi)
            entry["beacons_seen"] += 1
            entry["last_seen"] = now

            if enable_persistance:
                save_to_db(bssid=bssid)

        elif pkt.haslayer(Dot11ProbeReq):
            client = dot11.addr2.upper()
            ssid_raw = pkt[Dot11Elt].info if pkt.haslayer(Dot11Elt) else b""
            ssid = ssid_raw.decode('utf-8', errors='ignore').strip()
            if ssid:
                aps[client]["probes"].add(ssid)
                aps[client]["rssi_last"] = rssi

                if enable_persistance:
                    save_to_db(client_mac=client)

        elif dot11.type == 2 and dot11.subtype in [0, 8, 0x88]:
            if dot11.addr1 != "ff:ff:ff:ff:ff:ff":
                ap = None
                client = None
                if dot11.addr1 in aps:
                    ap = dot11.addr1
                    client = dot11.addr2.upper()
                    aps[ap]["clients"].add(client)
                elif dot11.addr2 in aps:
                    ap = dot11.addr2
                    client = dot11.addr1.upper()
                    aps[ap]["clients"].add(client)
                if ap and client:
                    cursor.execute('INSERT OR IGNORE INTO clients (mac, probes, rssi_last) VALUES (?, ?, ?)', (client, "", -100))
                    if enable_persistance:
                        save_to_db(bssid=ap)  # Updates associations

    except Exception as e:
        print(f"[WARN] Skipped bad packet: {e}")  # Log and continue


def show_info():
    print("\n\n=== Summary (APs sorted by strongest RSSI) ===")
    ap_list = [(bssid, data) for bssid, data in aps.items() if data["beacons_seen"] > 0]
    for bssid, data in sorted(ap_list, key=lambda x: x[1]["rssi_max"], reverse=True):
        clients_count = len(data["clients"])
        probes_str = ", ".join(data["probes"])[:60] + "..." if data["probes"] else ""
        print(f"{bssid} | {data['ssid']:25} | Ch:{data['channel']} ({data['freq']} MHz) | {data['encryption']} | RSSI max:{data['rssi_max']} dBm | Beacons:{data['beacons_seen']} | Clients:{clients_count} | Probes:{probes_str}")

    print("\n=== Probe Clients (no beacons) ===")
    for mac, data in aps.items():
        if data["beacons_seen"] == 0 and data["probes"]:
            print(f"{mac} probed: {', '.join(data['probes'])} (RSSI last: {data['rssi_last']})")


def main():
    print(f"Starting monitor on {IFACE} with channel hopping (interval {HOP_INTERVAL}s) and SQLite persistence to {DB_FILE}... Ctrl+C to stop")

    if HOPPING:
        hopper_thread = threading.Thread(target=channel_hopper, daemon=True)
        hopper_thread.start()

    try:
        sniff(iface=IFACE, prn=packet_handler, store=False, monitor=True)

    except KeyboardInterrupt:
        print("\n\n=== Summary (APs sorted by strongest RSSI) ===")
        ap_list = [(bssid, data) for bssid, data in aps.items() if data["beacons_seen"] > 0]
        for bssid, data in sorted(ap_list, key=lambda x: x[1]["rssi_max"], reverse=True):
            clients_count = len(data["clients"])
            probes_str = ", ".join(data["probes"])[:60] + "..." if data["probes"] else ""
            print(f"{bssid} | {data['ssid']:25} | Ch:{data['channel']} ({data['freq']} MHz) | {data['encryption']} | RSSI max:{data['rssi_max']} dBm | Beacons:{data['beacons_seen']} | Clients:{clients_count} | Probes:{probes_str}")

        print("\n=== Probe Clients (no beacons) ===")
        for mac, data in aps.items():
            if data["beacons_seen"] == 0 and data["probes"]:
                print(f"{mac} probed: {', '.join(data['probes'])} (RSSI last: {data['rssi_last']})")

    finally:
        if not enable_persistance:
            show_info()
        else:
            conn.close()



if __name__ == "__main__":
    main()

File: networking_tool/network_monitoring/network_scanner.go
package main

import (
	"fmt"  // TODO replace with log
	"log"
	"os"
	"os/exec"
	"os/signal"
	"sort"
	"strings"
	"sync"
	"syscall"
	"time"
	"database/sql"

	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
	_ "modernc.org/sqlite" // registers "sqlite" driver
	"gopkg.in/yaml.v2"
)


type Config struct {
	NetworkScanning struct {
		Interface     string  `yaml:"interface"`
		EnableHopping bool    `yaml:"enable_hopping"`
		HopInterval   float64 `yaml:"hop_interval"`
		Ch24G         []int   `yaml:"ch_2_4G"`
		Ch5G          []int   `yaml:"ch_5G"`
	} `yaml:"network_scanning"`
	Database struct {
		Path   string `yaml:"path"`
		Enable bool   `yaml:"enable"`
	} `yaml:"database"`
}


type APEntry struct {
	SSID        string
	Channel     string
	Freq        string
	Encryption  string
	RSSILast    int
	RSSIMax     int
	BeaconsSeen int
	LastSeen    float64
	Clients     map[string]bool
	Probes      map[string]bool
}


var (
	config        Config
	allChannels   []int
	aps           sync.Map  // map[string]*APEntry
	db            *sql.DB
	enablePersist bool
	commitInterval = 10 * time.Second
	hopping        bool
)



// Load config
func loadConfig() {
	config_file := "config/config.yaml"  // TODO CONST

	data, err := os.ReadFile(config_file)
	if err != nil {
		panic(fmt.Sprintf("Failed to read config: %v", err))
	}
	if err := yaml.Unmarshal(data, &config); err != nil {
		panic(fmt.Sprintf("Failed to parse config: %v", err))
	}
	allChannels = append(config.NetworkScanning.Ch24G, config.NetworkScanning.Ch5G...)  // TODO validate (non-empty, unique, sorted)
	hopping = config.NetworkScanning.EnableHopping
	enablePersist = config.Database.Enable
}


// Init DB
func initDB() {
	var err error

	db, err = sql.Open("sqlite", config.Database.Path)
	if err != nil {
		panic(fmt.Sprintf("Failed to open DB: %v", err))
	}

	// Create table: aps
	_, err = db.Exec(`CREATE TABLE IF NOT EXISTS aps (
		bssid TEXT PRIMARY KEY,
		ssid TEXT,
		channel TEXT,
		freq TEXT,
		encryption TEXT,
		rssi_last INTEGER,
		rssi_max INTEGER,
		beacons_seen INTEGER,
		last_seen REAL
	)`)
	if err != nil {
		panic(err)
	}

	// Create table: clients
	_, err = db.Exec(`CREATE TABLE IF NOT EXISTS clients (
		mac TEXT PRIMARY KEY,
		probes TEXT,
		rssi_last INTEGER
	)`)
	if err != nil {
		panic(err)
	}

	// Create table: associations
	_, err = db.Exec(`CREATE TABLE IF NOT EXISTS associations (
		ap_bssid TEXT,
		client_mac TEXT,
		PRIMARY KEY (ap_bssid, client_mac)
	)`)
	if err != nil {
		panic(err)
	}
}


// Batch save
func saveToDB() {
	if !enablePersist {
		return
	}

	tx, err := db.Begin()
	if err != nil {
		fmt.Printf("[DB] Begin failed: %v\n", err)
		return
	}
	defer tx.Rollback() // safety net

	stmtAP, _ := tx.Prepare("INSERT OR REPLACE INTO aps VALUES (?,?,?,?,?,?,?,?,?)")
	defer stmtAP.Close()

	stmtClient, _ := tx.Prepare("INSERT OR REPLACE INTO clients VALUES (?,?,?)")
	defer stmtClient.Close()

	stmtAssoc, _ := tx.Prepare("INSERT OR IGNORE INTO associations VALUES (?,?)")
	defer stmtAssoc.Close()

	stmtClientEnsure, _ := tx.Prepare("INSERT OR IGNORE INTO clients VALUES (?,?,?)")
	defer stmtClientEnsure.Close()

	aps.Range(func(key, value interface{}) bool {
		bssid := key.(string)
		e := value.(*APEntry)

		stmtAP.Exec(bssid, e.SSID, e.Channel, e.Freq, e.Encryption,
			e.RSSILast, e.RSSIMax, e.BeaconsSeen, e.LastSeen)

		probes := strings.Join(keys(e.Probes), ",")
		stmtClient.Exec(bssid, probes, e.RSSILast)

		for cl := range e.Clients {
			stmtAssoc.Exec(bssid, cl)
			stmtClientEnsure.Exec(cl, "", -100)
		}

		return true
	})

	if err := tx.Commit(); err != nil {
		fmt.Printf("[DB] Commit failed: %v\n", err)
	}
}


func keys(m map[string]bool) []string {
	ks := make([]string, 0, len(m))
	for k := range m {
		ks = append(ks, k)
	}
	return ks
}


// Channel hopper
func channelHopper() {
    for hopping {
        for _, ch := range allChannels {
            cmd := exec.Command("iw", "dev", config.NetworkScanning.Interface, "set", "channel", fmt.Sprint(ch))
            if err := cmd.Run(); err != nil {
                fmt.Printf("[HOP] Failed channel %d: %v\n", ch, err)
                time.Sleep(2 * time.Second) // backoff
                continue
            }
            time.Sleep(time.Duration(config.NetworkScanning.HopInterval * float64(time.Second)))
        }
    }
}


// Parse encryption from first few IEs (simplified)
func parseEncryption(ie *layers.Dot11InformationElement) string {
	for ie != nil {
		if ie.ID == 48 { // RSN → WPA2/WPA3
			return "WPA2/WPA3"
		}
		if ie.ID == 221 && len(ie.Info) >= 4 && ie.Info[0] == 0x00 && ie.Info[1] == 0x50 && ie.Info[2] == 0xF2 && ie.Info[3] == 0x01 {
			return "WPA"
		}
		// Move to next IE
		if len(ie.Payload) < 2 {
			break
		}
		nextLen := int(ie.Payload[1])
		if len(ie.Payload) < 2+nextLen {
			break
		}
		nextIE := &layers.Dot11InformationElement{}
		err := nextIE.DecodeFromBytes(ie.Payload[2:2+nextLen], gopacket.NilDecodeFeedback)
		if err != nil {
			break
		}
		ie = nextIE
	}
	return "OPEN"
}


// Packet handler
func handlePacket(packet gopacket.Packet) {
	dot11Layer := packet.Layer(layers.LayerTypeDot11)
	if dot11Layer == nil {
		return
	}
	dot11 := dot11Layer.(*layers.Dot11)

	radioLayer := packet.Layer(layers.LayerTypeRadioTap)
	rssi := -100
	if radioLayer != nil {
		rt := radioLayer.(*layers.RadioTap)
		if rt.DBMAntennaSignal != 0 {
			rssi = int(rt.DBMAntennaSignal)
		}
	}

	now := float64(time.Now().Unix())

	if dot11.Type == layers.Dot11TypeMgmtBeacon {
		bssid := dot11.Address3.String()
		ssid := "<hidden>"
		channel := "?"
		freq := "?"
		encryption := "OPEN"

		// Parse IEs manually from payload
		ie := &layers.Dot11InformationElement{}
		payload := dot11Layer.(*layers.Dot11).Payload
		if len(payload) > 0 {
			err := ie.DecodeFromBytes(payload, gopacket.NilDecodeFeedback)
			if err == nil {
				for ie != nil {
					if ie.ID == 0 { // SSID
						// Trim null bytes and invalid UTF-8
						ssidBytes := ie.Info
						for i, b := range ssidBytes {
							if b == 0 {
								ssidBytes = ssidBytes[:i]
								break
							}
						}
						ssid = strings.TrimSpace(string(ssidBytes))
						if ssid == "" {
							ssid = "<hidden>"
						}
						break // no need to keep walking after SSID
					}
					// Next IE
					if len(ie.Payload) < 2 {
						break
					}
					nextLen := int(ie.Payload[1])
					if len(ie.Payload) < 2+nextLen {
						break
					}
					nextIE := &layers.Dot11InformationElement{}
					err = nextIE.DecodeFromBytes(ie.Payload[2:2+nextLen], gopacket.NilDecodeFeedback)
					if err != nil {
						break
					}
					ie = nextIE
				}
			}
		}

		// Override channel from RadioTap if available
		if radioLayer != nil {
			rt := radioLayer.(*layers.RadioTap)
			if rt.ChannelFrequency != 0 {
				freq = fmt.Sprintf("%d", rt.ChannelFrequency)
				ch := freqToChannel(uint16(rt.ChannelFrequency))
				if ch != "?" {
					channel = ch
				}
			}
		}

		value, loaded := aps.LoadOrStore(bssid, &APEntry{
			Clients: make(map[string]bool),
			Probes:  make(map[string]bool),
		})
		entry := value.(*APEntry)
		if !loaded {
			entry.Clients = make(map[string]bool)
			entry.Probes = make(map[string]bool)
		}
		entry.SSID = ssid
		entry.Channel = channel
		entry.Freq = freq
		entry.Encryption = encryption
		entry.RSSILast = rssi
		if rssi < 0 && rssi > entry.RSSIMax {
			entry.RSSIMax = rssi
		}
		entry.BeaconsSeen++
		entry.LastSeen = now

	} else if dot11.Type == layers.Dot11TypeMgmtProbeReq {
		client := dot11.Address2.String()
		ssid := ""
		payload := dot11.Payload
		if len(payload) > 0 {
			ie := &layers.Dot11InformationElement{}
			err := ie.DecodeFromBytes(payload, gopacket.NilDecodeFeedback)
			if err == nil {
				for ie != nil {
					if ie.ID == 0 {
						ssid = string(ie.Info)
						break
					}
					if len(ie.Payload) < 2 {
						break
					}
					nextLen := int(ie.Payload[1])
					if len(ie.Payload) < 2+nextLen {
						break
					}
					nextIE := &layers.Dot11InformationElement{}
					err = nextIE.DecodeFromBytes(ie.Payload[2:2+nextLen], gopacket.NilDecodeFeedback)
					if err != nil {
						break
					}
					ie = nextIE
				}
			}
		}
		if ssid != "" {
			value, _ := aps.LoadOrStore(client, &APEntry{
				Clients: make(map[string]bool),
				Probes:  make(map[string]bool),
			})
			entry := value.(*APEntry)
			entry.Probes[ssid] = true
			entry.RSSILast = rssi
		}
	} else if dot11.Type.MainType() == layers.Dot11TypeData {
		if dot11.Address1.String() != "ff:ff:ff:ff:ff:ff" {
			ap := ""
			client := ""
			if _, ok := aps.Load(dot11.Address1.String()); ok {
				ap = dot11.Address1.String()
				client = dot11.Address2.String()
			} else if _, ok := aps.Load(dot11.Address2.String()); ok {
				ap = dot11.Address2.String()
				client = dot11.Address1.String()
			}
			if ap != "" && client != "" {
				value, _ := aps.Load(ap)
				entry := value.(*APEntry)
				entry.Clients[client] = true
			}
		}
	}
}


func freqToChannel(freq uint16) string {
	if freq >= 2400 && freq <= 2500 {
		return fmt.Sprintf("%d", (int(freq)-2407)/5)
	} else if freq >= 5000 && freq <= 6000 {
		return fmt.Sprintf("%d", (int(freq)-5000)/5)
	}
	return "?"
}


// Batch committer
func dbCommitter() {
	ticker := time.NewTicker(commitInterval)
	for range ticker.C {
		if enablePersist {
			saveToDB()
		}
	}
}


// Summary print (unchanged from previous)
func printSummary() {
	var apList []struct {
		bssid string
		entry *APEntry
	}
	aps.Range(func(key, value interface{}) bool {
		bssid := key.(string)
		entry := value.(*APEntry)
		if entry.BeaconsSeen > 0 {
			apList = append(apList, struct {
				bssid string
				entry *APEntry
			}{bssid, entry})
		}
		return true
	})
	sort.Slice(apList, func(i, j int) bool {
		return apList[i].entry.RSSIMax > apList[j].entry.RSSIMax
	})

	fmt.Println("\n=== Summary (APs sorted by strongest RSSI) ===")
	for _, item := range apList {
		e := item.entry
		clientsCount := len(e.Clients)
		probesStr := strings.Join(keys(e.Probes), ", ")
		if len(probesStr) > 60 {
			probesStr = probesStr[:60] + "..."
		}
		fmt.Printf("%s | %-25s | Ch:%s (%s MHz) | %s | RSSI max:%d dBm | Beacons:%d | Clients:%d | Probes:%s\n",
			item.bssid, e.SSID, e.Channel, e.Freq, e.Encryption, e.RSSIMax, e.BeaconsSeen, clientsCount, probesStr)
	}

	fmt.Println("\n=== Probe Clients (no beacons) ===")
	aps.Range(func(key, value interface{}) bool {
		mac := key.(string)
		e := value.(*APEntry)
		if e.BeaconsSeen == 0 && len(e.Probes) > 0 {
			probesStr := strings.Join(keys(e.Probes), ", ")
			fmt.Printf("%s probed: %s (RSSI last: %d)\n", mac, probesStr, e.RSSILast)
		}
		return true
	})
}


func main() {
    loadConfig()
    log.Printf("Config loaded | Interface: %s | Hopping: %v | Persistence: %v",
        config.NetworkScanning.Interface, hopping, enablePersist)

    if enablePersist {
        initDB()
        defer db.Close()
        log.Printf("Database opened: %s", config.Database.Path)
    }

    if hopping {
        go channelHopper()
    }

    go dbCommitter()

    // === Pcap setup ===
    handle, err := pcap.OpenLive(config.NetworkScanning.Interface, 65536, true, pcap.BlockForever)
    if err != nil {
        log.Fatalf("pcap.OpenLive failed: %v", err)
    }
    defer handle.Close()

    if err := handle.SetBPFFilter("type mgt or type ctl"); err != nil {
        log.Fatalf("SetBPFFilter failed: %v", err)
    }

    packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
    log.Println("Listening for 802.11 management/control packets... (Ctrl+C to stop)")

    // Packet loop in goroutine
    go func() {
        for packet := range packetSource.Packets() {
            handlePacket(packet)
        }
    }()

    // Wait for Ctrl+C
    sigChan := make(chan os.Signal, 1)
    signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
    <-sigChan

    log.Println("Shutting down...")
    printSummary()
    if enablePersist {
        saveToDB() // final batch
        log.Println("Final DB save completed")
    }
}
